RavenDB-22566 - Cannot cancel a restore of a snapshot

grisha-kotler opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-22566/Cannot-cancel-a-restore-of-a-snapshot

### Additional description

Port from 6.0: https://github.com/ravendb/ravendb/pull/18785
- Allow to cancel the restore when downloading the snapshot.
- Get the file size of the snapshot from the API.
- Add the progress of downloading the restore file.

### Type of change

- [x] Bug fix
- [ ] Regression bug fix
- [x] Optimization
- [ ] New feature

### How risky is the change?

- [x] Low 
- [ ] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [x] It has been verified by manual testing

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

grisha-kotler committed:
RavenDB-22566 - get the snapshot size from the request and not from the stream length which might not be supported
- add progress

grisha-kotler committed:
RavenDB-22566 - cannot cancel the download of a snapshot from cloud services

grisha-kotler committed:
RavenDB-22566 - address PR comments

grisha-kotler committed:
RavenDB-22566 - update the free space that is needed to restore the snapshot

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/GoogleCloud/RavenGoogleCloudClient.cs:
```diff
@@ -150,6 +150,15 @@ public Task<Object> GetObjectAsync(string fileName)
 
+        public async Task<Size> GetObjectSizeAsync(string fileName)
+        {
+            var obj = await GetObjectAsync(fileName);
```
Leaving a comment for future regarding this API method: `Fetches the information about an object asynchronously. This does not retrieve the content of the object itself.`

grisha-kotler replied:
done

ppekrol commented on src/Voron/Impl/Backup/StreamExtensions.cs:
```diff
@@ -31,5 +32,17 @@ public static void CopyTo(this Stream source, Stream destination, Action<int> on
 
+        public static async Task CopyToAsync(this Stream source, Stream destination, Action<int> onProgress, CancellationToken cancellationToken)
+        {
+            var readBuffer = new byte[DefaultBufferSize];
```
`tryf` + `ArrayPool` ?

grisha-kotler replied:
done

grisha-kotler committed:
RavenDB-22566 - use ArrayPool

grisha-kotler committed:
RavenDB-22566 - add a comment for the method

grisha-kotler committed:
RavenDB-22566 - use ArrayPool instead of thread static
The restore process isn't a dedicated thread so it won't ever be cleaned by calling ThreadLocalCleanup.Run().

grisha-kotler committed:
RavenDB-22566 - use DefaultBufferSize
